pkgs/applications/misc/dmenu-rs/default.nix

```diff
@@ -0,0 +1,113 @@
+{ lib
+, fetchFromGitHub
+, rustPlatform
+, clang
+, libclang
+, llvm
+, m4
+, pkg-config
+, expat
+, fontconfig
+, libX11
+, libXft
+, libXinerama
+, python3
+}:
+
+# The dmenu-rs package has extensive plugin support. However, this derivation
+# only builds the package with the default set of plugins. If you'd like to
+# further customize dmenu-rs you can build it from the source.
+# See: https://github.com/Shizcow/dmenu-rs#plugins
+rustPlatform.buildRustPackage rec {
+  pname = "dmenu-rs";
+  version = "5.5.1";
+
+  src = fetchFromGitHub {
+    owner = "Shizcow";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-WpDqBjIZ5ESnoRtWZmvm+gNTLKqxL4IibRVCj0yRIFM=";
+  };
+
+  nativeBuildInputs = [
+    clang
+    libclang
+    llvm
+    m4
+    pkg-config
+    python3
+  ];
+
+  buildInputs = [
+    expat
+    fontconfig
+    libX11
+    libXft
+    libXinerama
+  ];
+
+  # The dmenu-rs repository does not include a Cargo.lock because of its
+  # dynamic build and plugin support. Generating it with make and checking it
+  # in to nixpkgs here was the easiest way to supply it to buildRustPackage.
+  # See: https://github.com/Shizcow/dmenu-rs/issues/34#issuecomment-757415584
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
+
+  # Several nixpkgs' functions from its rust build tooling expect cargo files
+  # to exist at the project's root directory unless otherwise specified.
+  cargoRoot = "src";
+
+  # The rust-xcb dependency dynamically generates rust code at build time.
+  # This derivation uses nixpkgs rust functions that vendor each cargo
+  # dependency's source code into the READ-ONLY nix store. To avoid the code
+  # generation step failing, we copy the rust-xcb source out of the nix store
+  # and make it writeable.
+  # See: https://github.com/rust-x-bindings/rust-xcb/tree/v0.8.2
+  patchPhase = ''
+    chmod +w "$NIX_BUILD_TOP/cargo-vendor-dir"
+    mkdir -p "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    cp -r "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2/." "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    unlink "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+    mv "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite" "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+    chmod -R +w "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2/"
```

Comment:
why do you need the copy step, and not only the recursive chmod?

Reply:
`.../cargo-vendor-dir/xcb-0.8.2/` is a symlink to the nix store. I don't think I can `chmod` the nix store inside of the build and in general making anything in the nix store writeable seems conceptually wrong?

Reply:
oh it's originally a symlink, I thought it was a copy. My bad.

you can remove the final chmod -R by using cp --no-preserve=mod then